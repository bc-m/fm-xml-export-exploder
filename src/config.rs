#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum CatalogType {
    Accounts,
    BaseDirectory,
    BaseTable,
    CalcsForCustomFunctions,
    CustomFunctions,
    CustomMenu,
    CustomMenuSet,
    ExternalDataSource,
    ExtendedPrivileges,
    FieldsForTables,
    FileAccess,
    Layout,
    Library,
    OptionsForValueLists,
    PrivilegeSets,
    Relationship,
    Script,
    StepsForScripts,
    TableOccurrence,
    Theme,
    ValueList,
}

impl CatalogType {
    pub fn from_bytes(bytes: &[u8]) -> Option<Self> {
        match bytes {
            b"AccountsCatalog" => Some(Self::Accounts),
            b"BaseDirectoryCatalog" => Some(Self::BaseDirectory),
            b"BaseTableCatalog" => Some(Self::BaseTable),
            b"CalcsForCustomFunctions" => Some(Self::CalcsForCustomFunctions),
            b"CustomFunctionsCatalog" => Some(Self::CustomFunctions),
            b"CustomMenuCatalog" => Some(Self::CustomMenu),
            b"CustomMenuSetCatalog" => Some(Self::CustomMenuSet),
            b"ExternalDataSourceCatalog" => Some(Self::ExternalDataSource),
            b"ExtendedPrivilegesCatalog" => Some(Self::ExtendedPrivileges),
            b"FieldsForTables" => Some(Self::FieldsForTables),
            b"FileAccessCatalog" => Some(Self::FileAccess),
            b"LayoutCatalog" => Some(Self::Layout),
            b"LibraryCatalog" => Some(Self::Library),
            b"OptionsForValueLists" => Some(Self::OptionsForValueLists),
            b"PrivilegeSetsCatalog" => Some(Self::PrivilegeSets),
            b"RelationshipCatalog" => Some(Self::Relationship),
            b"ScriptCatalog" => Some(Self::Script),
            b"StepsForScripts" => Some(Self::StepsForScripts),
            b"TableOccurrenceCatalog" => Some(Self::TableOccurrence),
            b"ThemeCatalog" => Some(Self::Theme),
            b"ValueListCatalog" => Some(Self::ValueList),
            _ => None,
        }
    }

    pub fn get_config(&self) -> CatalogConfig {
        match self {
            Self::Accounts => CatalogConfig {
                catalog_item_name: b"Account".to_vec(),
                out_folder_name: "accounts".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "Account/AccountReference/@id".to_string(),
            },
            Self::BaseDirectory => CatalogConfig {
                catalog_item_name: b"BaseDirectory".to_vec(),
                out_folder_name: "base_directory".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::BaseTable => CatalogConfig {
                catalog_item_name: b"BaseTable".to_vec(),
                out_folder_name: "table_stubs".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::CalcsForCustomFunctions => CatalogConfig {
                catalog_item_name: b"CustomFunctionCalc".to_vec(),
                out_folder_name: "custom_function_calcs".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "CustomFunctionCalc/CustomFunctionReference/@id".to_string(),
            },
            Self::CustomFunctions => CatalogConfig {
                catalog_item_name: b"CustomFunction".to_vec(),
                out_folder_name: "custom_function_stubs".to_string(),
                wrapped_in_object_list: true,
                uses_folders: true,
                id_path: "".to_string(),
            },
            Self::CustomMenu => CatalogConfig {
                catalog_item_name: b"CustomMenu".to_vec(),
                out_folder_name: "custom_menus".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::CustomMenuSet => CatalogConfig {
                catalog_item_name: b"CustomMenuSet".to_vec(),
                out_folder_name: "custom_menu_sets".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::ExternalDataSource => CatalogConfig {
                catalog_item_name: b"ExternalDataSource".to_vec(),
                out_folder_name: "external_data_sources".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::ExtendedPrivileges => CatalogConfig {
                catalog_item_name: b"ExtendedPrivilege".to_vec(),
                out_folder_name: "extended_privileges".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::FieldsForTables => CatalogConfig {
                catalog_item_name: b"FieldCatalog".to_vec(),
                out_folder_name: "table_fields".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "FieldCatalog/BaseTableReference/@id".to_string(),
            },
            Self::FileAccess => CatalogConfig {
                catalog_item_name: b"Authorization".to_vec(),
                out_folder_name: "file_access".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::Layout => CatalogConfig {
                catalog_item_name: b"Layout".to_vec(),
                out_folder_name: "layouts".to_string(),
                wrapped_in_object_list: false,
                uses_folders: true,
                id_path: "Layout/LayoutReference/@id".to_string(),
            },
            Self::Library => CatalogConfig {
                catalog_item_name: b"BinaryData".to_vec(),
                out_folder_name: "libraries".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "BinaryData/LibraryReference/@id".to_string(),
            },
            Self::OptionsForValueLists => CatalogConfig {
                catalog_item_name: b"ValueList".to_vec(),
                out_folder_name: "value_list_options".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "ValueList/ValueListReference/@id".to_string(),
            },
            Self::PrivilegeSets => CatalogConfig {
                catalog_item_name: b"PrivilegeSet".to_vec(),
                out_folder_name: "privilege_sets".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::Relationship => CatalogConfig {
                catalog_item_name: b"Relationship".to_vec(),
                out_folder_name: "relationships".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::Script => CatalogConfig {
                catalog_item_name: b"Script".to_vec(),
                out_folder_name: "script_stubs".to_string(),
                wrapped_in_object_list: false,
                uses_folders: true,
                id_path: "".to_string(),
            },
            Self::StepsForScripts => CatalogConfig {
                catalog_item_name: b"Script".to_vec(),
                out_folder_name: "script_steps".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "Script/ScriptReference/@id".to_string(),
            },
            Self::TableOccurrence => CatalogConfig {
                catalog_item_name: b"TableOccurrence".to_vec(),
                out_folder_name: "table_occurrences".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::Theme => CatalogConfig {
                catalog_item_name: b"Theme".to_vec(),
                out_folder_name: "themes".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::ValueList => CatalogConfig {
                catalog_item_name: b"ValueList".to_vec(),
                out_folder_name: "value_list_stubs".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
        }
    }
}

#[derive(Debug)]
pub struct CatalogConfig {
    pub catalog_item_name: Vec<u8>,
    pub out_folder_name: String,
    pub wrapped_in_object_list: bool,
    pub uses_folders: bool,
    pub id_path: String,
}

pub struct Flags {
    pub parse_all_lines: bool,
    pub lossless: bool,
    pub verbose: Option<Option<String>>,
}
