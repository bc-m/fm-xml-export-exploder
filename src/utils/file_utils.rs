use anyhow::{anyhow, Error, Result};
use std::path::PathBuf;

/// Validate that a path exists and is a directory
pub fn valid_dir_or_throw(dir_path: &PathBuf) -> Result<(), Error> {
    let metadata = std::fs::metadata(dir_path)
        .map_err(|_| anyhow!("Path '{}' not exists", dir_path.display()))?;

    match metadata.is_dir() {
        true => Ok(()),
        false => Err(anyhow!("Path '{}' is not a directory", dir_path.display())),
    }
}

/// Join scope ID and name into a formatted string
pub fn join_scope_id_and_name(scope_id: &str, scope_name: &str) -> String {
    format!("{} - ID {}", scope_name, scope_id)
}

/// Escape filename by replacing special characters with underscores
pub fn escape_filename(filename: &str) -> String {
    let escaped_filename = filename
        .chars()
        .map(|c| match c {
            '/' | '\\' | ':' | '*' | '?' | '"' | '<' | '>' | '|' | '{' => '_',
            _ => c,
        })
        .collect::<String>();

    escaped_filename.trim().to_string()
}

/// Check if a line should be skipped based on content
pub fn should_skip_line(line: &str) -> bool {
    if line.contains("<TagList></TagList>") {
        return true;
    };
    if line.contains("<OwnerID></OwnerID>") {
        return true;
    };
    if line.contains("<Options>0</Options>") {
        return true;
    }; // Default
    if line.contains("<Options>1048576</Options>") {
        return true;
    }; // Breakpoint
    if line.contains("<UUID>") && line.contains("</UUID>") {
        return true;
    };

    false
}
