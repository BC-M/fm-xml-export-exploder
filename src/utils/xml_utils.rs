use crate::utils::attributes::get_attributes;
use quick_xml::events::{BytesCData, BytesEnd, BytesStart, BytesText, Event};
use quick_xml::Reader;
use std::io::{BufRead, Read};

pub fn start_element_to_string(e: &BytesStart) -> String {
    let mut complete_tag = format!("<{}", local_name_to_string(e.name().as_ref()));
    for attr in get_attributes(e).unwrap() {
        // de-noise TODO: make optional
        match attr.0.as_str() {
            "nextvalue" | "UUID" | "index" => {
                continue;
            }
            _ => {}
        }

        complete_tag.push_str(&format!(" {}=\"{}\"", attr.0, attr.1));
    }
    complete_tag.push('>');

    complete_tag
}

pub fn cdata_element_to_string(e: &BytesCData) -> String {
    let mut content = String::new();
    content.push_str("<![CDATA[");
    content.push_str(cdata_to_string(e).as_str());
    content.push_str("]]>");

    content
}

pub fn text_element_to_string(e: &BytesText, escape: bool) -> String {
    if escape {
        decode_xml_special_characters(text_to_string(e))
    } else {
        text_to_string(e)
    }
}

pub fn end_element_to_string(e: &BytesEnd) -> String {
    let element_name = local_name_to_string(e.name().as_ref());
    format!("</{}>", element_name)
}

pub fn local_name_to_string(local_name: &[u8]) -> String {
    return match std::str::from_utf8(local_name) {
        Ok(text) => text.to_string(),
        Err(_) => String::new(),
    };
}

pub fn text_to_string(e: &BytesText) -> String {
    return match e.unescape() {
        Ok(text) => text.to_string(),
        Err(_) => String::new(),
    };
}

pub fn cdata_to_string(e: &BytesCData) -> String {
    return match std::str::from_utf8(e) {
        Ok(text) => text.to_string(),
        Err(_) => String::new(),
    };
}

fn decode_xml_special_characters(input: String) -> String {
    input
        .replace('&', "&amp;")
        .replace('\"', "&quot;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
        .replace('\'', "&apos;")
        .replace('\r', "&#13;")
}

pub fn skip_element<R: Read + BufRead>(reader: &mut Reader<R>, _: &BytesStart) {
    let mut depth = 1;
    let mut buf: Vec<u8> = Vec::new();
    loop {
        match reader.read_event_into(&mut buf) {
            Err(_) => continue,
            Ok(Event::Eof) => break,
            Ok(Event::Start(_)) => {
                depth += 1;
            }
            Ok(Event::End(_)) => {
                depth -= 1;
                if depth == 0 {
                    break;
                }
            }
            _ => {}
        }
        buf.clear();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_decode_xml_special_characters() {
        assert_eq!(
            decode_xml_special_characters("This & that \"test\" <tag>".to_string()),
            "This &amp; that &quot;test&quot; &lt;tag&gt;"
        );
    }
}
