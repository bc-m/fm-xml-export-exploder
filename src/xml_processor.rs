use std::io::{BufRead, BufReader, Read};
use std::path::{Path, PathBuf};
use std::{fs::File, time::Instant};

use anyhow::{bail, Context, Error, Result};
use encoding_rs_io::DecodeReaderBytes;
use quick_xml::events::{BytesEnd, BytesStart, Event};
use quick_xml::reader::Reader;

use crate::catalog::xml_explode_catalog;
use crate::config::{CatalogType, Flags};
use crate::script_sanitizer::create_sanitized_scripts;
use crate::supporting::process_supporting_element;
use crate::utils::attributes::get_attribute;
use crate::utils::xml_utils::{end_element_to_string, start_element_to_string, XmlEventType};
use crate::utils::{add_fm_name_to_base_path, build_out_dir_path, write_xml_file, FolderStructure};
use crate::utils::{push_line_to_skeleton, version_string_to_number};
use crate::Skeleton;

/// Context for XML catalog processing
pub struct ProcessingContext<'a, R: Read + BufRead> {
    pub reader: &'a mut Reader<R>,
    pub path_stack: &'a mut Vec<Vec<u8>>,
    pub current_out_dir: PathBuf,
    pub skeleton: &'a mut Skeleton,
    pub flags: &'a Flags,
}

/// Process a single XML file and explode it into individual files
pub fn explode_xml(
    fm_export_file_path: &PathBuf,
    root_out_dir: &Path,
    flags: &Flags,
) -> Result<(), Error> {
    let start = Instant::now();
    let fm_export_file_name = fm_export_file_path.file_name().unwrap().to_str().unwrap();

    // Open XML file
    let file = File::open(fm_export_file_path)
        .with_context(|| format!("Error opening file {}", fm_export_file_path.display(),))?;

    // Initialize variables
    let mut fm_file_name = String::new();
    let mut saxml_version = String::new();
    let mut depth = 0;
    let mut cf_folder_structure: Option<FolderStructure> = None;
    let mut script_folder_structure: Option<FolderStructure> = None;
    let mut buf = Vec::new();

    // Instantiate processing context which will be passed around to various functions
    let mut context = ProcessingContext {
        reader: &mut Reader::from_reader(BufReader::new(DecodeReaderBytes::new(file))),
        path_stack: &mut Vec::new(),
        current_out_dir: PathBuf::new(),
        skeleton: &mut Skeleton::default(),
        flags,
    };

    // Iterate over XML events
    loop {
        match context.reader.read_event_into(&mut buf) {
            Err(e) => {
                println!("Error in {}: {}", fm_export_file_path.display(), e);
                break;
            }
            Ok(Event::Eof) => break,
            Ok(Event::Start(start_tag)) => {
                // context.start_tag = Some(&e);
                // if depth <= 4 {
                //     println!("📟 {}:{} {}", depth, "  ".repeat(depth), start_element_to_string(&e, flags));
                // }
                context.path_stack.push(start_tag.name().as_ref().to_vec());
                push_start_to_skeleton(
                    &start_tag,
                    context.skeleton,
                    context.path_stack,
                    context.flags,
                );
                match depth {
                    0 => {
                        process_root_element(&start_tag, &mut fm_file_name, &mut saxml_version)?;
                        context.current_out_dir =
                            add_fm_name_to_base_path(root_out_dir, &fm_file_name, &saxml_version);
                        build_out_dir_path(&mut context, "", true);
                    }
                    1 => {
                        let do_continue = process_supporting_elements(&mut context, &start_tag);
                        if do_continue {
                            continue;
                        }
                    }
                    3 => {
                        let is_supported_catalog = process_catalog_elements(
                            &mut context,
                            &start_tag,
                            &saxml_version,
                            &mut cf_folder_structure,
                            &mut script_folder_structure,
                        )?;
                        if is_supported_catalog {
                            continue;
                        }
                    }
                    _ => {}
                }
                depth += 1;
            }
            Ok(Event::End(e)) => {
                context.path_stack.pop();
                push_end_to_skeleton(&e, context.skeleton, context.path_stack, context.flags);
                depth -= 1;
                // if depth <= 4 {
                //     println!("🚩 {}:{} {}", depth, "  ".repeat(depth), end_element_to_string(&e));
                // }
            }
            _ => {}
        }
        buf.clear()
    }

    // Write skeleton file if in lossless mode
    if flags.lossless {
        write_skeleton_file(&context.current_out_dir, context.skeleton, context.flags)?;
    }

    println!(
        "→ {} finished in {} ms.",
        fm_export_file_name,
        start.elapsed().as_millis()
    );

    Ok(())
}

fn process_root_element(
    e: &BytesStart,
    fm_file_name: &mut String,
    saxml_version: &mut String,
) -> Result<(), Error> {
    match e.name().as_ref() {
        b"FMDynamicTemplate" | b"FMSaveAsXML" => {
            *fm_file_name = get_attribute(e, "File")
                .unwrap()
                .strip_suffix(".fmp12")
                .unwrap()
                .to_string();
            *saxml_version = get_attribute(e, "version").unwrap().to_string();
        }
        _ => {
            bail!("Unsupported XML-format");
        }
    }
    Ok(())
}

/// Process supporting elements (Metadata, DDR_INFO) that are not part of the main structure
fn process_supporting_elements<R: Read + BufRead>(
    context: &mut ProcessingContext<'_, R>,
    start_tag: &BytesStart,
) -> bool {
    let mut do_continue = false;
    match start_tag.name().as_ref() {
        b"Metadata" => {
            process_supporting_element(context, start_tag, "metadata");
            do_continue = true;
        }
        b"DDR_INFO" => {
            process_supporting_element(context, start_tag, "ddr_info");
            do_continue = true;
        }
        _ => {}
    }
    do_continue
}

fn process_catalog_elements<R: Read + BufRead>(
    context: &mut ProcessingContext<'_, R>,
    start_tag: &BytesStart,
    saxml_version: &str,
    cf_folder_structure: &mut Option<FolderStructure>,
    script_folder_structure: &mut Option<FolderStructure>,
) -> Result<bool, Error> {
    let catalog_type = match CatalogType::from_bytes(start_tag.name().as_ref()) {
        Some(catalog_type) => catalog_type,
        None => {
            println!(
                "⚠️ Skipping unsupported catalog: {}",
                std::str::from_utf8(start_tag.name().as_ref()).unwrap()
            );
            return Ok(false);
        }
    };

    // Detect catalog type
    let mut catalog_config = catalog_type.get_config();
    // println!("             ✅ Processing catalog: {}, {}", start_element_to_string(e, flags), catalog_config.out_folder_name);

    // Handle special cases that need folder structures
    let folder_structure = match catalog_type {
        CatalogType::CalcsForCustomFunctions => cf_folder_structure.as_ref(),
        CatalogType::StepsForScripts => script_folder_structure.as_ref(),
        _ => None,
    };

    // Handle ValueList version-specific behavior
    if catalog_type == CatalogType::ValueList
        && version_string_to_number(saxml_version) < version_string_to_number("2.2.2.0")
    {
        catalog_config.out_folder_name = "value_list_options".to_string();
    }

    let xml_out_dir_path = build_out_dir_path(context, &catalog_config.out_folder_name, false);
    let base_dir_path = context.current_out_dir.clone();
    context.current_out_dir = xml_out_dir_path.clone();

    let folder_structure_result =
        xml_explode_catalog(context, start_tag, folder_structure, &catalog_config);

    context.current_out_dir = base_dir_path; // Restore to its original value

    // Update folder structures for catalogs that return them
    if let Some(folder_structure) = folder_structure_result {
        match catalog_type {
            CatalogType::CustomFunctions => *cf_folder_structure = Some(folder_structure),
            CatalogType::Script => *script_folder_structure = Some(folder_structure),
            _ => {}
        }
    }

    // Handle post-processing for StepsForScripts
    if catalog_type == CatalogType::StepsForScripts {
        let sanitized_scripts_dir_path = build_out_dir_path(context, "script_sanitized", false);
        create_sanitized_scripts(
            &xml_out_dir_path,
            &sanitized_scripts_dir_path,
            script_folder_structure.as_ref(),
            context.flags,
        );
    }
    Ok(true)
}

fn write_skeleton_file(
    skeleton_dir_path: &Path,
    skeleton: &Skeleton,
    flags: &Flags,
) -> Result<(), Error> {
    let skeleton_file_path = skeleton_dir_path.join("skeleton.xml");
    write_xml_file(&skeleton_file_path, &skeleton.content, 0, flags);
    Ok(())
}

fn push_start_to_skeleton(
    e: &BytesStart,
    skeleton: &mut Skeleton,
    path_stack: &[Vec<u8>],
    flags: &Flags,
) {
    if flags.lossless {
        match path_stack.len() {
            1..=3 => {
                push_line_to_skeleton(
                    skeleton,
                    path_stack.len(),
                    1,
                    start_element_to_string(e, flags).as_str(),
                    true,
                    XmlEventType::Start,
                );
            }
            4 => {
                if let Some(last) = path_stack.iter().rev().nth(1) {
                    let parent = std::str::from_utf8(last).unwrap();
                    if ["AddAction", "ModifyAction"].contains(&parent) {
                        push_line_to_skeleton(
                            skeleton,
                            path_stack.len(),
                            1,
                            start_element_to_string(e, flags).as_str(),
                            true,
                            XmlEventType::Start,
                        );
                    }
                }
            }
            5 => {
                if let Some(last) = path_stack.iter().rev().nth(2) {
                    let grandparent = std::str::from_utf8(last).unwrap();
                    if ["AddAction", "ModifyAction"].contains(&grandparent) {
                        push_line_to_skeleton(
                            skeleton,
                            path_stack.len(),
                            1,
                            start_element_to_string(e, flags).as_str(),
                            true,
                            XmlEventType::Start,
                        );
                    }
                }
            }
            _ => {}
        }
    }
}

fn push_end_to_skeleton(
    e: &BytesEnd,
    skeleton: &mut Skeleton,
    path_stack: &[Vec<u8>],
    flags: &Flags,
) {
    if flags.lossless {
        match path_stack.len() {
            0..=2 => {
                push_line_to_skeleton(
                    skeleton,
                    path_stack.len(),
                    1,
                    end_element_to_string(e).as_str(),
                    false,
                    XmlEventType::End,
                );
            }
            3 => {
                if let Some(last) = path_stack.iter().next_back() {
                    let parent = std::str::from_utf8(last).unwrap();
                    if ["AddAction", "ModifyAction"].contains(&parent) {
                        push_line_to_skeleton(
                            skeleton,
                            path_stack.len(),
                            1,
                            end_element_to_string(e).as_str(),
                            false,
                            XmlEventType::End,
                        );
                    }
                }
            }
            4 => {
                if let Some(last) = path_stack.iter().rev().nth(1) {
                    let grandparent = std::str::from_utf8(last).unwrap();
                    if ["AddAction", "ModifyAction"].contains(&grandparent) {
                        push_line_to_skeleton(
                            skeleton,
                            path_stack.len(),
                            1,
                            end_element_to_string(e).as_str(),
                            false,
                            XmlEventType::End,
                        );
                    }
                }
            }
            _ => {}
        }
    }
}
