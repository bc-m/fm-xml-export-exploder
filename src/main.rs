use std::{fs, path::PathBuf, time::Instant};

use anyhow::Result;
use clap::Parser;
use rayon::prelude::*;

use crate::config::Flags;
use crate::utils::file_utils::valid_dir_or_throw;
use crate::utils::xml_utils::XmlEventType;
use crate::xml_processor::explode_xml;

mod catalog;
mod config;
mod script_sanitizer;
mod script_steps;
mod supporting;
mod tests;
mod utils;
mod xml_processor;

/// Parse all as XML exported FileMaker solutions from source directory and explode them to target directory.
#[derive(Parser)]
#[command(version, about, long_about = None)]
struct Args {
    /// The source directory to read input
    source: PathBuf,

    /// The target directory to write output
    target: PathBuf,

    /// Parse all lines (or skip less important ones to reduce noise)
    #[arg(short, long)]
    all_lines: bool,

    /// Retain all information from the main xml (or skip less important catalogs and attributes)
    #[arg(short, long)]
    lossless: bool,

    /// Enable debug/verbose output. Optionally filter by catalog name.
    /// Example: --verbose "CustomFunctionsCatalog"
    #[arg(short, long, value_name = "FILTER", num_args(0..=1))]
    verbose: Option<Option<String>>,
}

#[derive(Debug, Default)]
pub struct Skeleton {
    pub content: String,
    pub previous_line: String,
    pub previous_event_type: XmlEventType,
}

fn main() -> Result<()> {
    let start = Instant::now();

    let args = Args::parse();
    let in_dir = args.source;
    let out_dir = args.target;
    let flags = Flags {
        parse_all_lines: args.all_lines,
        lossless: args.lossless,
        verbose: args.verbose,
    };

    match &flags.verbose {
        Some(Some(filter)) => println!("🔍 Verbose mode (filtered on {})", filter),
        Some(None) => println!("🔊 Verbose mode (all)"),
        None => (), // silent
    };

    valid_dir_or_throw(&in_dir)?;

    // Read directory contents
    let paths = fs::read_dir(in_dir)?
        .filter_map(|entry| entry.ok().map(|e| e.path())) // Filter out directories and unwrap results
        .filter(|path| path.is_file() && path.extension().unwrap_or_default() == "xml") // Filter XML files
        .collect::<Vec<_>>(); // Collect paths into a vector

    println!("Start processing {} files...", paths.len());

    // Process XML files in parallel
    paths.par_iter().for_each(|path| {
        match explode_xml(path, &out_dir, &flags) {
            Ok(_) => {}
            Err(err) => {
                let file_name = path.file_name().unwrap().to_str().unwrap();
                eprintln!("Failed to process file '{}': {}", file_name, err)
            }
        };
    });

    let duration = start.elapsed();
    if duration.as_secs() > 9 {
        println!("Completed in {:?} seconds.", duration.as_secs());
    } else {
        println!("Completed in {:?} ms.", duration.as_millis());
    }

    Ok(())
}
