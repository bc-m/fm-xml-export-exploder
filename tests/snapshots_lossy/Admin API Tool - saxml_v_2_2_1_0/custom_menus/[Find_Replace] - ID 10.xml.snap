---
source: src/main.rs
expression: output_content
---
<CustomMenu name="[Find/Replace]" id="10">
	<UUID modifications="0" userName="Mike Duncan" accountName="Admin" timestamp="2020-12-18T12:17:11">8266B03D-DF03-474E-B933-C6D418154E92</UUID>
	<MenuItemList membercount="0"></MenuItemList>
	<Conditions>
		<Install>
			<Calculation><![CDATA[1]]></Calculation>
			<ChunkList hash="A93F17BC54CA1958073B692697D5ED21">
				<Chunk type="NoRef">1</Chunk>
			</ChunkList>
		</Install>
	</Conditions>
	<Comment></Comment>
	<Base name="Suchen/Ersetzen" value="149"></Base>
	<Options browseMode="True" findMode="True" previewMode="True"></Options>
</CustomMenu>
